File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.cc
#include <iostream>
#include <algorithm>
using namespace std;
int n,s,arr[25],ans;

void func(int sum, int idx){
	if(idx == n){
		if(sum == s) ans++;
		return;
	}
	
	func(sum + arr[idx], idx + 1);
	func(sum, idx + 1);
}

int main() {
	cin >> n >> s;
	for(int i = 0; i < n; i++) cin >> arr[i];
	sort(arr, arr + n);
	
	func(0,0);
	if(s == 0) ans--;
	cout << ans;
	return 0;
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.io.*;
import java.util.*;

class Main {
	public static int n,s,ans;
	public static int[] arr;
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		s = Integer.parseInt(st.nextToken());
		
		StringTokenizer st2 = new StringTokenizer(br.readLine());
		
		arr = new int[n];
		for(int i = 0; i < n; i++){
			arr[i] = Integer.parseInt(st2.nextToken());
		}
		
		go(0,0);
		if(s == 0) ans--;
		System.out.println(ans);
	}
	
	public static void go(int idx, int sum){
		if(idx == n) {
			if(sum == s) ans++;
			return;
		}
		go(idx + 1, sum);
		go(idx + 1, sum + arr[idx]);
	}
}